P3663: R3 (changes from Tomasz and LWG review) (#3)
mhoemmen/wg21-proposals: submdspan_future/submdspan-future.md

>  Any object of type `s` is *collapsing slice*.

Reviewer: ```suggestion

```

Did I do that?  Is this a left-over copy-paste?

* [3.1]{.pnum} $S_k$ models `convertible_to<index_type>`,

* [3.2]{.pnum} $S_k$ models _`index-pair-like`_`<index_type>`,

* [3.3]{.pnum} `is_convertible_v<` $S_k$ `, full_extent_t>` is `true`, or

* [3.4]{.pnum} $S_k$ is a specialization of `strided_slice`.

[4]{.pnum} *Preconditions*: For each rank index $k$ of `extents()`, all of the following are `true`:

* [4.1]{.pnum} if $S_k$ is a specialization of `strided_slice`, $s_k$`.extent` is equal to zero or $s_k$`.stride` is greater than zero; and

* [4.2]{.pnum} 0 &le; _`first_`_`<index_type,`$k$`>(slices...)` <br>
    0 &le; _`last_`_`<`$k$`>(extents(), slices...)` <br>
    0 &le; _`extents().extent(`$k$`)`
:::

::: add
[3]{.pnum} *Mandates*: For each rank index $k$ of `src`, `SliceSpecifiers...[`$k$`]` is a valid `submdspan` slice type for the $k^{th}$ extent of `Extents`, and

[4]{.pnum} *Preconditions*: For each rank index $k$ of `src`, `slices...[`$k$`]` is a valid slice for the $k^{th}$ extent of `src`.
:::

[5]{.pnum} Let `sub_ext` be the result of `submdspan_extents(extents(), slices...)` and let `SubExtents` be `decltype(sub_ext)`.

> [6]{.pnum} Let `sub_strides` be an `array<SubExtents::index_type, SubExtents::rank()>` such that for each rank index $k$ of `extents()` for which [_`map-rank`_`[`$k$`]` is not `dynamic_extent`]{.rm}[`slices...[`$k$`]` is not collapsing slice]{.add}, `sub_strides[`[_`map-rank`_`[`$k$`]`]{.rm}[$MAP\_RANK$`(slices, `$k$`)`]{.add}`]` equals:

Reviewer: ```suggestion
[6]{.pnum} Let `sub_strides` be an `array<SubExtents::index_type, SubExtents::rank()>` such that for each rank index $k$ of `extents()` for which [_`map-rank`_`[`$k$`]` is not `dynamic_extent`]{.rm}[`slices...[`$k$`]` is not a collapsing slice]{.add}, `sub_strides[`[_`map-rank`_`[`$k$`]`]{.rm}[$MAP\_RANK$`(slices, `$k$`)`]{.add}`]` equals:

```

```

::: add

```
template<class T>
concept @_is-strided-slice_@ = _see below_;
```

> [1]{.pnum} The concept _`is-strided-slice`_`<T>` is satisfied and modleded if `T` is specialization of `strided_slice`.

Reviewer: ```suggestion
[1]{.pnum} The concept _`is-strided-slice`_`<T>` is satisfied and modeled if `T` is a specialization of `strided_slice`.

```

```
template<class IndexType, class S>
constexpr auto @_canonical-index_@(S s);
```

> [2]{.pnum} *Mandates*: If `S` models _`integral-constant-like`_ and if, then `extents<IndexType>::`_`index-cast`_`(S::value)` is 

Reviewer: It looks like the "and if" before the comma is superfluous.

```suggestion
[2]{.pnum} *Mandates*: If `S` models _`integral-constant-like`_, then `extents<IndexType>::`_`index-cast`_`(S::value)` is 

```

representable as a value of type `IndexType`.

[3]{.pnum} *Preconditions*: `extents<IndexType>::`_`index_cast`_`(std::move(s))` is representable as a value of type `IndexType`.

[4]{.pnum} Effects: Equivalent to:

* [4.1]{.pnum} `cw<IndexType(extents<IndexType>::`_`index_cast`_`(S::value))>` if `S` models _`integral-constant-like`_;

* [4.2]{.pnum} `IndexType(extents<IndexType>::`_`index-cast`_`(std::move(s)))` otherwise.

```
template<class IndexType, class S>
constexpr auto @_canonical-slice_@(S s);
```

> [5]{.pnum} *Mandates*: `S` is a `submdspan` slice type for `IndexType` ([mdspan.sub.slices]) 

Reviewer: ```suggestion
[5]{.pnum} *Mandates*: `S` is a `submdspan` slice type for `IndexType` ([mdspan.sub.slices]).

```

[6]{.pnum} *Effects*: Equivalent to:
```
if constexpr (is_convertible_v<S, full_extent_t>)
{

>   return static_cast<full_extent_t>(std::move(s));

Reviewer: What happens if `S` is actually a const reference?  Then we wouldn't be able to move from it.

Author: The function takes them by value.

}
else if constexpr (is_convertible_v<S, IndexType>)
{
  return @_canonical-index_@<IndexType>(std::move(s));
}
else if constexpr (@_is_strided_slice_@<S>)
{
  return strided_slice{
           @_canonical-index_@<IndexType>(std::move(s.offset)),
           @_canonical-index_@<IndexType>(std::move(s.extent)),
           @_canonical-index_@<IndexType>(std::move(s.stride)),
         };
}
else
{
  auto [f, l] = std::move(s);
  auto cf =  @_canonical-index_@<IndexType>(std::move(f));
  auto cl =  @_canonical-index_@<IndexType>(std::move(l));
  return strided_slice{cf, cl, @_canonical-index_@<IndexType>(cl-cf)};
}
```

:::

```c++
template<class IndexType, size_t k, class... SliceSpecifiers>
  constexpr IndexType @_first_@_(SliceSpecifiers... slices);
```
// TODO add extents parameter, make non variadic

::: add
[*Editorial note*: Former paragraph 1 has been renumbered to 5. -- *end note*]

:::

::: rm

[5]{.pnum} *Mandates*: `IndexType` is a signed or unsigned integer type.

[6]{.pnum} Let $ϕ_k$ denote the following value:

* [6.1]{.pnum} $s_k$ if $S_k$ models `convertible_to<IndexType>`;

> * [6.2]{.pnum} otherwise, `get<0>(`$s_k$`)` if $S_k$ models _`index-pair-like`_`<IndexType>`;

Reviewer: Note to self, to check whether $\phi_k$ can ever see _`index-pair-like`_.

Reviewer: Note to self: This doesn't matter, because $\phi_k$ is going away.

* [6.3]{.pnum} otherwise, _`de-ice`_`(`$s_k$`.offset)` if $S_k$ is a specialization of `strided_slice`;

* [6.4]{.pnum} otherwise, `0`.

[7]{.pnum} *Preconditions*: $ϕ_k$ is representable as a value of type `IndexType`.

[8]{.pnum} *Returns*: `extents<IndexType>::`_`index-cast`_`(`$ϕ_k$`)`

:::

::: add

[5]{.pnum} *Mandates*: `SliceSpecifiers...[k]` is canonical `k`^{th} `mdspan` slice type of `Extents`.

> [8]{.pnum} *Returns*: l, where [l, u) is `submdspan` slice range of `slices...[k]` for `k` dimension of `src`.

Reviewer: ```suggestion
[8]{.pnum} *Returns*: $L$, where $[L, U)$ is the `submdspan` slice range of `slices...[k]` for the `k`^{th} dimension of `src`.

```

(It's really hard to tell lowercase "l" from 1.)

:::

```
template<size_t k, class Extents, class... SliceSpecifiers>
  constexpr auto @_last_@_(const Extents& src, SliceSpecifiers... slices);
```
// TODO make non variadic

::: rm
[9]{.pnum} *Mandates*: `Extents` is a specialization of `extents`.

[10]{.pnum} Let `index_type` be `typename Extents::index_type`.

[11]{.pnum} Let $λ_k$ denote the following value:

* [11.1]{.pnum} _`de-ice`_`(`$s_k$`) + 1` if $S_k$ models `convertible_to<index_type>`.

* [11.2]{.pnum} `get<1>(`$s_k$`)` if $S_k$ models _`index-pair-like`_`<index_type>`; otherwise

* [11.3]{.pnum} _`de-ice`_`(`$s_k$`.offset) + `_`de-ice`_`(`$s_k$`.extent)` if $S_k$ is a specialization of `strided_slice`; otherwise

* [11.4]{.pnum} `src.extent(k)`.

[12]{.pnum} *Preconditions*: $λ_k$ is representable as a value of `type index_type`.

[13]{.pnum} *Returns*: `Extents​::`_`​index-cast`_`(`$λ_k$`)`.

:::

::: add

> [0]{.pnum} *Mandates*: `SliceSpecifiers...[k]` is canonical `mdspan` slice type for `k` dimension of `Extents`.

Reviewer: ```suggestion
[0]{.pnum} *Mandates*: `SliceSpecifiers...[k]` is a canonical `mdspan` slice type for the `k`^{th} extent of `Extents`.

```

* [5.1]{.pnum} `cw<IndexType(extents<IndexType>::`_`index_cast`_`(S::value))>` if `S` models _`integral-constant-like`_;

* [5.2]{.pnum} `IndexType(extents<IndexType>::`_`index-cast`_`(std::move(s)))` otherwise.

```
template<class IndexType, class S>
constexpr auto @_canonical-slice_@(S s);
```

[6]{.pnum} *Effects*: Equivalent to:
```
if constexpr (is_convertible_v<S, full_extent_t>)
{
  return static_cast<full_extent_t>(std::move(s));
}
else if constexpr (is_convertible_v<S, IndexType>)
{
  return @_canonical-index_@<IndexType>(std::move(s));
}
else if constexpr (@_is_strided_slice_@<S>)
{
  return strided_slice{
           @_canonical-index_@<IndexType>(std::move(s.offset)),
           @_canonical-index_@<IndexType>(std::move(s.extent)),
           @_canonical-index_@<IndexType>(std::move(s.stride)),
         };
}
else
{
  auto [f, l] = std::move(s);
  auto cf =  @_canonical-index_@<IndexType>(std::move(f));
  auto cl =  @_canonical-index_@<IndexType>(std::move(l));
  return strided_slice{cf, cl, @_canonical-index_@<IndexType>(cl-cf)};
}

```c++
template<class IndexType, size_t k, class... SliceSpecifiers>
  constexpr IndexType @_first_@_(SliceSpecifiers... slices);
```

> // TODO: We should replace it with just

Author: We do not need mandates and cosntrains. This is internal function, that is always called in-contract, so just returns

    * [1.4.2]{.pnum} `is_convertible_v<decltype(std::move(L))>, IndexType>` is `true`, and

    * [1.4.3]{.pnum} `is_convertible_v<decltype(std::move(U))>, IndexType>` is `true`.

[2]{.pnum} Given a signed or unsigned integer `IndexType`, a type $S$ is a *canonical `submdspan` index type for `IndexType`* if $S$ is either `IndexType` or `constant_wrapper<v, IndexType>` for some `v` of type `IndexType`, such that `v` is greater than or equal to zero.

[3]{.pnum} Given a signed or unsigned integer `IndexType`, a type $S$ is a *canonical `submdspan` slice type for `IndexType`* if exactly one of the following is `true`:

* [3.1]{.pnum} $S$ is `full_extent_t`;

* [3.2]{.pnum} $S$ is a canonical `submdspan` index type for `IndexType`; or

* [3.3]{.pnum} $S$ is a specialization of `strided_slice` where all of the following hold:

    * [3.3.1]{.pnum} $S$`::offset_type`, $S$`::extent_type`, and $S$`::stride_type` are all canonical `submdspan` index types for `IndexType`; and

    * [3.3.2]{.pnum} if $S$`::stride_type` and $S$`::extent_type` are both specializations of  `constant_wrapper`,
      then $S$::`stride_type::value` is greater than zero.

> [4]{.pnum} Let `IndexType` be a signed or unsigned integer type.  A type `S` is a *collapsing slice type* if `S` is a canonical `submdspan` index type for `IndexType`.  An object `s` of type `S` is a *collapsing slice* if `S` is a collapsing slice type.

Author: I do not really like the introduction of `collapsing slice type` and `unit slice type`, we do not use them anywhere so they are just necessary.  We should apply Occam's razor here, and remove there.

I would much prefer saying:
> Let `IndexType` be a signed or unsigned integer type , an object `s` is a *collapsing slice* if it's type ....

We could also free this definition from index type, if we say that it is *collapsing slice* if it's not specialization of strided_slice or full_extent.

[13]{.pnum} *Returns*: `Extents​::`_`​index-cast`_`(`$λ_k$`)`.

::: add
```
enum class @_check-static-bounds-result_@ {
  @_in-bounds_@,
  @_out-of-bounds_@,
  @_unknown_@
};

template<size_t k, class Slice, class IndexType, size_t... Exts>
  constexpr @_check-static-bounds-result_@ @_check-static-bounds_@(
    const extents<IndexType, Exts...>&);
```

> [14]{.pnum} *Returns*:

Author: I just put the whole things into words, about the ranges being subrange of each other.

* [4.2]{.pnum} interval of `k` dimension of `e`  is not know statically, or

* [4.2]{.pnum} interval of `k` dimension of `e` contains slice range of `s` for `k` dimension of `e`.

[5]{.pnum} Given an object `e` of type `E` that is specialization of `extents`,
an object `s` of type `S`, **`s` is valid slice for `k` dimension of `e`** if:

* [5.1]{.pnum} `S` is valid slice type for `k` dimension of `E`, and

* [5.2]{.pnum} interval of `k` dimension of `e` contains slice range of `s` for `k` dimension of `e`.

[6]{.pnum} Given an object `e` of type `E` that is specialization of `extents`,
 an pack of expression `slcs`, **`slcs` are valid slices for `e`** if:

* [6.1]{.pnum} `sizeof...(slcs)` is equal to `E::rank()`,

* [6.2]{.pnum} for each rank index `r` of `e`, the `slcs...[r]` is valid slice for `r` dimension of `e`.

```
template<class IndexType, size_t... Extents, class... Slices>
constexpr auto submdspan_canonicalize_slices(
  const extents<IndexType, Extents...>& src, Slices... raw_slices);
```

Let `slices` denote pach _`canonical-slice`_`<IndexType>(raw_slices)`,

[8]{.pnum} *Constraints*: `sizeof...(raw_slices)` equals `sizeof...(Extents)`.

> [9]{.pnum} *Mandates*: For each rank index $k$ of `src`, $slices...[k]$ is valid expression and `decltype(slices...[k])` is a valid slice type for `k` dimension of `src`.

Author: The `canonical-slice` is valid expressions, brinds the Mandates about it being an acceptable slice type. I.e. the function is just ill-formed otherwise.

```
template<class IndexType, class... Extents, class... SliceSpecifiers>
  constexpr auto submdspan_extents(const extents<IndexType, Extents...>& src,
                                   SliceSpecifiers... raw_slices);
```

[2]{.pnum} Let `slices` be the pack resulting from the following statement.

```
auto [...slices] = submdspan_canonicalize_slices(src, raw_slices...).
```

[1]{.pnum} *Constraints*: `sizeof...(SlicesSpecifiers)` equals `Extents::rank()`.

[2]{.pnum} *Mandates*: submdspan_canonicalize_slices(src, raw_slices...) is valid expressions. 
For each rank index $k$ of `src`, `decltype(slices...[k])` is a valid slice type for `k` dimension of `src`.

[3]{.pnum} *Preconditions*: For each rank index $k$ of `src`, $slices...[k]$ is a valid slice for `k` dimension of `src`.

> [4]{.pnum} Let `SubExtents` be a specialization of `extents` such that:

Author: We need to bring rest of the wording here to be able to rewrite it.

[2]{.pnum} *Constraints*: `sizeof...(slices)` equals `extents_type::rank()`.

::: rm

[3]{.pnum} *Mandates*: For each rank index $k$ of `extents()`, exactly one of the following is `true`:

* [3.1]{.pnum} $S_k$ models `convertible_to<index_type>`,

* [3.2]{.pnum} $S_k$ models _`index-pair-like`_`<index_type>`,

* [3.3]{.pnum} `is_convertible_v<` $S_k$ `, full_extent_t>` is `true`, or

* [3.4]{.pnum} $S_k$ is a specialization of `strided_slice`.

[4]{.pnum} *Preconditions*: For each rank index $k$ of `extents()`, all of the following are `true`:

* [4.1]{.pnum} if $S_k$ is a specialization of `strided_slice`, $s_k$`.extent` is equal to zero or $s_k$`.stride` is greater than zero; and

* [4.2]{.pnum} 0 &le; _`first_`_`<index_type,`$k$`>(slices...)` <br>
    0 &le; _`last_`_`<`$k$`>(extents(), slices...)` <br>

>     0 &le; _`extents().extent(`$k$`)`

Reviewer: Does `<br>` here mean "and"?

Author: Look at old line 1209. Question for you.

Reviewer: The original formatting was different!  https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2642r5.html , [mdspan.submdspan.mapping.common] 4.2.  (I'm using R5 because R6 is a giant image, because a certain person, neither of the two of us, decided that would be the fastest way to get red and green diffs for LWG review ; - ) .)  The original intent was "and."

Reviewer: In any case, we're deleting these lines, so I'm not too worried.

:::

::: add

[3]{.pnum} *Mandates*: For each rank index $k$ of `src`:

> * [3.1]{.pnum} `SliceSpecifiers...[k]` is `submdspan` slice type for `k` dimension of `Extents`, and

Reviewer: ```suggestion
* [3.1]{.pnum} `SliceSpecifiers...[k]` is a `submdspan` slice type for the `k`^{th} extent of `Extents`, and

```

> * [3.2]{.pnum} `decltype(`_`canonical-slice`_`<IndexType>(slices...[k]))` is valid `submdspan` slice type for `k` dimension of `Extents`.

Reviewer: What's a "valid `submdspan` slice type"?  I thought we were just calling it "`submdspan` slice type."

```suggestion
* [3.2]{.pnum} `decltype(`_`canonical-slice`_`<IndexType>(slices...[k]))` is a `submdspan` slice type for the `k`^{th} extent of `Extents`.

```

> [4]{.pnum} *Preconditions*: For each rank index $k$ of `src`, _`canonical-slice`_`<IndexType>(slices...[k])` is a valid slice for `k` dimension of `src`.

Reviewer: ```suggestion
[4]{.pnum} *Preconditions*: For each rank index $k$ of `src`, _`canonical-slice`_`<IndexType>(slices...[k])` is a valid slice for the `k`^{th} extent of `src`.

```

[*Note 1*: This invocation of `submdspan_mapping` selects a function call via overload resolution on a candidate set that includes the lookup set found by argument-dependent lookup ([basic.lookup.argdep]). — *end note*]

[5]{.pnum} *Constraints*:

* [5.1]{.pnum} `sizeof...(slices)` equals `Extents​::​rank()`, and

* [5.2]{.pnum} [the expression `submdspan_mapping(src.mapping(),` `slices...)` is well-formed when treated as an unevaluated operand]{.rm}[`Layout::mapping<Extents>` models _`sliceable-mapping`_]{.add}.

> [3]{.pnum} *Preconditions*: For each rank index $k$ of `src`, $slices...[k]$ is a valid slice for `k` dimension of `src`.

Reviewer: ```suggestion
[3]{.pnum} *Preconditions*: For each rank index $k$ of `src`, $slices...[k]$ is a valid slice for the `k`^{th} extent of `src`.

```

> * [5.2]{.pnum} [the expression `submdspan_mapping(src.mapping(),` `slices...)` is well-formed when treated as an unevaluated operand]{.rm}[`Layout::mapping<Extents>` meets the requirements of _`mapping-sliceable-with-full-extents`_]{.add}.

Author: This should be models not meets requirements.

// This goes here.
[2]{.pnum} *Mandates*: submdspan_canonicalize_slices(src, raw_slices...) is valid expressions. 
For each rank index $k$ of `src`, `decltype(slices...[k])` is a valid slice type for `k` dimension of `src`.

[3]{.pnum} *Preconditions*: For each rank index $k$ of `src`, $slices...[k]$ is a valid slice for `k` dimension of `src`.

[6]{.pnum} *Mandates*:

> ::: rm

Author: We should keep the first two mandates, as they define behavior for particular slice types that we use.
